subsys/net/lib/fota_download/src/fota_download.c

```diff
@@ -7,22 +7,250 @@
 #include <stdint.h>
 #include <zephyr.h>
 #include <flash.h>
-#include <net/download_client.h>
 #include <dfu/mcuboot.h>
 #include <dfu/flash_img.h>
 #include <pm_config.h>
 #include <logging/log.h>
+#include <net/download_client.h>
 #include <net/fota_download.h>
+#include <nrf_socket.h>
+#include <net/socket.h>
 
 LOG_MODULE_REGISTER(fota_download, CONFIG_FOTA_DOWNLOAD_LOG_LEVEL);
 
+static struct	download_client		 dfu;
+static struct	flash_img_context	 flash_img;
 static		fota_download_callback_t callback;
-static struct	flash_img_context flash_img;
-static struct	download_client dfu;
+
+#ifdef CONFIG_FOTA_DOWNLOAD_MODEM_FW_UPDATE_SUPPORT
+static int			modem_dfu_socket_fd;
+static struct	k_sem		modem_fw_delete_sem;
+static struct	k_delayed_work  modem_fw_delete_dwork;
+static struct k_work_q modem_fw_delete_offload_work_q;
```

Comment:
If you see the sample it puts the fota_download task as a sys workqueue item. What I experienced is that since the code executing in fota_download is executed as part of the submitted work publishing work into the syswork queue would not be scheduled as the initial work has not finished. I guess it's implemented as FIFO and not really scheduled hence the name queue :) 

However this could be the wrong solution to the problem, hence it is still in a draft stage. Do you have a suggestion on what should be the best way of handling this? 

Should the whole fota_download be put into a thread by sample instead? Maybe we could avoid using work-queue at all in fota_download.